### PR TITLE
#1680 Arrows on curved edges

### PR DESCRIPTION
**Pull request by jacomyma**

### CODE REVIEW

**Review by mbastian — COMMENTED** [review]
Great work! Couple of remarks

1. When the radius is greater than the edge distance it creates weird circles
Screenshot 2022-09-02 at 12 06 53

2. Can you check that the edge label positions are now based on the arc edges (not bezier)? Not sure how it's done.
Screenshot 2022-09-02 at 12 15 58

File: modules/PreviewPlugin/src/main/java/org/gephi/preview/plugin/renderers/ArrowRenderer.java [hunk 1]
```diff
@@ -228,23 +235,55 @@ public Helper(
...
-            p2.mult(size * BASE_RATIO);
-            p2.add(p1r);
-
-            p3 = new Vector(direction.y, -direction.x);
-            p3.mult(size * BASE_RATIO);
-            p3.add(p1r);
+            if (properties.getBooleanValue(PreviewProperty.EDGE_CURVED)) {
+                // Change the direction to account for the curvature
+                Double newAngle = Math.atan2(direction.y, direction.x);
+                Double curvature = (double)properties.getFloatValue(ARC_CURVENESS);
```
Comment (mbastian) [inline modules/PreviewPlugin/src/main/java/org/gephi/preview/plugin/renderers/ArrowRenderer.java:244]: Can use `properties.getDoubleValue()` instead

File: modules/PreviewPlugin/src/main/java/org/gephi/preview/plugin/renderers/EdgeRenderer.java [hunk 2]
```diff
@@ -103,6 +105,7 @@ public class EdgeRenderer implements Renderer {
     protected EdgeColor defaultColor = new EdgeColor(EdgeColor.Mode.MIXED);
     protected boolean defaultEdgeCurved = true;
     protected float defaultBezierCurviness = 0.2f;
```
Comment (mbastian) [inline modules/PreviewPlugin/src/main/java/org/gephi/preview/plugin/renderers/EdgeRenderer.java:107]: To be removed as not needed

File: modules/PreviewPlugin/src/main/java/org/gephi/preview/plugin/renderers/EdgeRenderer.java [hunk 3]
```diff
@@ -185,6 +188,11 @@ public void preProcess(PreviewModel previewModel) {
             properties.putValue(BEZIER_CURVENESS, defaultBezierCurviness);
```
Comment (mbastian) [inline modules/PreviewPlugin/src/main/java/org/gephi/preview/plugin/renderers/EdgeRenderer.java:188]: Same, to remove

File: modules/PreviewPlugin/src/main/java/org/gephi/preview/plugin/renderers/EdgeRenderer.java [hunk 4]
```diff
@@ -84,6 +85,7 @@ public class EdgeRenderer implements Renderer {
     public static final String EDGE_MIN_WEIGHT = "edge.min-weight";
     public static final String EDGE_MAX_WEIGHT = "edge.max-weight";
     public static final String BEZIER_CURVENESS = "edge.bezier-curveness";
```
Comment (mbastian) [inline modules/PreviewPlugin/src/main/java/org/gephi/preview/plugin/renderers/EdgeRenderer.java:87]: We should add a `@Deprecated` javadoc comment on this one

File: modules/PreviewPlugin/src/main/java/org/gephi/preview/plugin/renderers/EdgeRenderer.java [hunk 5]
```diff
@@ -628,6 +663,58 @@ public Helper(
 
                 v1 = computeCtrlPoint(x1, y1, direction, factor, n);
                 v2 = computeCtrlPoint(x2, y2, direction, -factor, n);
+
+                // Arc radius
+                r = (double) (length / properties.getFloatValue(ARC_CURVENESS));
```
Comment (mbastian) [inline modules/PreviewPlugin/src/main/java/org/gephi/preview/plugin/renderers/EdgeRenderer.java:668]: Same, use `getDoubleValue()`

File: modules/PreviewPlugin/src/main/java/org/gephi/preview/plugin/renderers/EdgeRenderer.java [hunk 6]
```diff
@@ -628,6 +663,58 @@ public Helper(
...
+                bbx = xc-r;
+                bby = yc-r;
+                bbw = 2*r;
+                bbh = 2*r;
+                astart = -180*(angle1)/Math.PI;
+                asweep = (180*(angle1-angle2)/Math.PI+720)%360 - 360;
+            }
+
+            private Double computeTheThing(Double radius_curvature_edge, Double truncature_length) {
+                // I don't want to explain what this is. I'll try though.
```
Comment (mbastian) [inline modules/PreviewPlugin/src/main/java/org/gephi/preview/plugin/renderers/EdgeRenderer.java:708]: We want to stay away from any personal comments in the code as it's not professional

**Comment by jacomyma** [comment]
Also to do: edit the presets to replace the Bézier curve radius by the arc curve radius (different property name). @mbastian could you point to where it is in the code?

**Pull request merged**